README.md
Memberpages Plugin for CMSimple_XH
==================================

Due to lack of time [svasti](http://svasti.de/) is currently unable to maintain
the Memberpages plugin. As *provisional* *solution* it has therefore been decided
to put the plugin here, so that the CMSimple_XH developers are able to do the
necessary maintenance (at least do some odd fixes), and release new versions
from time to time.

[![stable: 3.6.7](https://img.shields.io/badge/stable-3.6.7-green.svg)](https://github.com/cmsimple-xh/memberpages/releases/tag/3.6.7)
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)


For further information see help/help.htm.


changelog.txt
C H A N G E L O G memberpages Plugin for CMSimple(_XH))

=============================================
Series 3 by Svasti (svasti.de)
---------------------------------------------
3.6.7 Jan 2023
    - see <https://github.com/cmsimple-xh/memberpages/milestone/6?closed=1>
3.6.6 Jan 2023
    - see <https://github.com/cmsimple-xh/memberpages/milestone/5?closed=1>
3.6.5 Sep 2017
    - see <https://github.com/cmsimple-xh/memberpages/milestone/4?closed=1>
3.6.4 Aug 2017
    - see <https://github.com/cmsimple-xh/memberpages/milestone/3?closed=1>
3.6.3 Aug 2017
    - see <https://github.com/cmsimple-xh/memberpages/milestone/2?closed=1>
3.6.2 Apr 2017
    - see <https://github.com/cmsimple-xh/memberpages/milestone/1?closed=1>
3.6.1 Feb 2016
    - Small bugfix (reported by Ludwig)
3.6 Feb 2016
    - Changed an anonymous function call which worked only in php 5.3 to 'create_function' 
      which also works in php 5.2 (Solution by Holger)
    - Added possibility to hide pages on login (i.e. the login page), suggested by Volkmar
    - Wrong input in login field will return to the login field (via anchor) with error notice,
      useful for one page solutions, suggested by Ludwig
    - removed all settings for page on error (entry of false username or password) as this seems 
      rather troublesome when a member made a typing error
    - removed all "tag"-function calls
    - works with PHP 7 and XH 1.7
    - added option for members notice after login on top of optionally every page or only member pages
    - Autologin now as checkbox (with noscript handling), when autologin enabled no logout button
    - Register me and submit button get bigger on small screens
    - Security requirements for new password entered by the user can be set to different degrees
    - New Slovak language file by Tata

3.5.2 August 2015
    - Additional access-setting: VIP access pages with equal or lower levels
      while non-VIPs can only access pages with same or no level (requested by mhz)
    - Bugfix (VIP function in non-German sites was affected)
    - Some wordings improved
    - Ready for php 7 (needed a fix in the memberlist.php from cmb and a fix in the password generator)

3.5.1 Apr 2015
    - bug fix in plugin activation program  (thx cmb)
    - bug fix in config labeling expiration text (the label had partly the same wording as the text
      so that the regex changed not only the text but the label too)

3.5 Apr 2015
    - Added CMSIMPLE_URL definition for use with older XH or CMSimple 4.x versions
    - Code clean up, code put into functions
    - Special welcome page possible for members with higher level
    - "Register me" added, sends email to webmaster, collected data will be presented at the bottom 
      of the members list with import function and send-password-to-user function.
    - revamped interactive panels: password forgotten, register me, change data, with extensive answers
    - Autologin button now after login. Clicking autologin will be displayed in the user info and will be
      still be displayed at the next visit
    - Scroll position is retained in back end members list view via short lived cookie
    - Password now hashed also in session var for better security
    - Hash algrorithm changed from md5 to sha256, which is safer although not as save as
      the hash function of newer php versions.
    - Put old functionality for CMSimple scripting memberpage call in an extra file only included if necessary
    - Added opcache_invalidate
    - Exclusive file locking added to saving different files
    - Now memberslist, logfile, register-me and logged-in data all with .htaccess default at 
      userfiles/plugins/memberpages
    - Members list is saved in ini format in a protected php file
    - No more "file"-dialog in memberpages backend, no more direct upload/download of members list 
      (could be done via ftp though)
    - Detailed update function with import of old members list and log file, checks if new file are there and
      writable, checks for save_mode=on, choice to keep old directory or use new default directory
    - Beefed up log view with size in bytes and entries and button to reduce log to last 500 entries
    - Length of random password can be set to 4, 5 or 6 characters,
      if members want to change their passwords, the same minimum number of char is required
    - New Password are required to have at least an uppercase letter, a lowercase letter and a cipher.
    - Expires date converts entries like 1.1.15 to 2015-01-01 (suggested by Ludwig)
      Expire date saved in unix time stamp. Configurable warning (days before membership expires)
      on log in of members.
    - Expired memberships will be highlighted in back end members list.
    - Error message in backend if any member is without password (suggested by Ludwig)
    - Bugfix (thx to cmb)
    - many spelling corrections contributed by Michael (mhz)
    - After passwords are changed by members in the members panel they are shown as dots in the members list
      and are not editable any more by the webmaster
    - Pluginactivation now in separate file, together with function to import old members list
    - Small security improvement in checking the entered email address (contributed by cmb)
    - memberslogin() and membersnotice() can take the argument 1 for smaller information of logged in members


3.4.3 Jan 2015
    - Bugfix (Bug reported by Wolfgang, permission data was lost when saving emails was unselected)

3.4.2 Oct 2014
    - Bugfix (in activation form) reported by cmb

3.4.1 April 2014
    - Fixed bug: Colered "M" in pagetabs didn't work due to renaming of a CMSimple_XH 1.6 variable from
      CMSimple_XH 1.6 beta to 1.6 final

3.4 (October 2013)
    - If expiration date or full name is deselected. values will not be lost (saved in hidden fields).
    - Now log entries start with date

admin.php
<?php

/**
 * Backend of Memberpages_XH
 *
 * Copyright 2012-2016 by svasti@svasti.de
 */

if ((!function_exists('sv')) || preg_match('!admin.php!i', sv('PHP_SELF'))) die('Access denied');

define('MEMBERPAGES_VERSION', '3.6.7');

include 'funcs.php';

if (function_exists('XH_registerPluginMenuItem')) {
    XH_registerPluginMenuItem('memberpages',$plugin_tx['memberpages']['menu_main'],
        '?&memberpages&admin=plugin_main&action=plugin_text');
    XH_registerPluginMenuItem('memberpages',$plugin_tx['memberpages']['menu_log'],
        '?&memberpages&admin=plugin_config&action=log');
    XH_registerStandardPluginMenuItems(false);
    XH_registerPluginMenuItem('memberpages',$plugin_tx['memberpages']['menu_credits'],
        '?&memberpages&admin=plugin_config&action=credits');
}

if (function_exists('XH_wantsPluginAdministration') && XH_wantsPluginAdministration('memberpages')
   || isset($memberpages) && $memberpages === 'true'
) {
    $plugin = basename(dirname(__FILE__),"/");
    $admin  = isset($_POST['admin']) ? $_POST['admin'] : $admin = isset($_GET['admin']) ? $_GET['admin'] : '';
    $action = isset($_POST['action']) ? $_POST['action'] : $action = isset($_GET['action']) ? $_GET['action'] : '';

    $memberpageimages=$pth['folder']['plugins'].$plugin."/images/";


    $plugin_main_on   = ( !$admin || $admin=='plugin_main') ?        ' class="membp_selected"':'';
    $config_on        = ($admin=='plugin_config' && $action=='plugin_edit')? ' class="membp_selected"':'';
    $stylesheet_on    = $admin=='plugin_stylesheet'?                 ' class="membp_selected"':'';
    $language_on      = $admin=='plugin_language'?                   ' class="membp_selected"':'';
    $credits_on       = $action=='credits'?                          ' class="membp_selected"':'';
    $log_on           = $action=='log'?                              ' class="membp_selected"':'';

    $o .= '<p class="membp_admin_menu">' . "\n"
       .  '<a'.$plugin_main_on  .' href="?&amp;' . $plugin . '&amp;admin=plugin_main&amp;action=plugin_text">'
       .  $plugin_tx[$plugin]['menu_main'].'</a>&nbsp; ' . "\n"
       .  '<a'.$log_on          .' href="?&amp;' . $plugin . '&amp;admin=plugin_config&amp;action=log">'
       .  $plugin_tx[$plugin]['menu_log'].'</a>&nbsp; ' . "\n"
       .  '<a'.$config_on       .' href="?&amp;' . $plugin . '&amp;admin=plugin_config&amp;action=plugin_edit">'
       .  $plugin_tx[$plugin]['menu_config'].'</a>&nbsp; ' . "\n"
       .  '<a'.$stylesheet_on   .' href="?&amp;' . $plugin . '&amp;admin=plugin_stylesheet&amp;action=plugin_text">'
       .  $plugin_tx[$plugin]['menu_css'].'</a>&nbsp; ' . "\n"
       .  '<a'.$language_on     .' href="?&amp;' . $plugin . '&amp;admin=plugin_language&amp;action=plugin_edit">'
       .  $plugin_tx[$plugin]['menu_language'].'</a>&nbsp; ' . "\n"
       .  '<a                      href="'.        $pth['file']['plugin_help'] . '" target="_blank">'
       .  $plugin_tx[$plugin]['menu_help'].'</a>&nbsp; ' . "\n"
       .  '<a'.$credits_on      .' href="?&amp;' . $plugin . '&amp;admin=plugin_config&amp;action=credits">'
       .  $plugin_tx[$plugin]['menu_credits'].'</a></p>' . "\n";

    if($admin == '' && $action != 'credits') $admin = 'plugin_main';

    if($admin != 'plugin_main' and $admin != 'plugin_config')
    {
        $o .= plugin_admin_common($action,$admin,$plugin);
    }

    if($action=='credits') {
        $o.= '<h2>Memberpages_XH ' . constant('MEMBERPAGES_VERSION') . '</h2>' . "\n"
           . '<p>&copy; 2012-2016 by <a href="http://frankziesing.de/cmsimple/">svasti</a></p>'."\n"
           . '<p>based on:</p>' ."\n" . '<ul>'
           . '<li>versions 3.0 and higher by <a href="http://frankziesing.de/cmsimple/">svasti</a></li>'."\n"
           . '<li>versions 2.0–2.3 (2009-2010) by <a href="http://www.ge-webdesign.de">ge-webdesign.de</a></li>'."\n"
           . '<li>versions 0.1–1.7 (2005-2007) by <a href="http://cmsimpleplugins.svarrer.dk">cmsimpleplugins.svarrer.dk</a></li>'."\n".'</ul>' . "\n"
           . '<p>This program is free software: you can redistribute it and/or modify '
	       . 'it under the terms of the GNU General Public License as published by '
	       . 'the Free Software Foundation, either version 3 of the License, or '
	       . '(at your option) any later version.</p>'
           . '<p>For a copy of the GNU General Public License see https://www.gnu.org/licenses/gpl.html</p>' ."\n"
           . '<p><small><b>Acknowledgement:</b></small><br />'
           . 'Memberpages_XH uses Memberlist_XH by <a href="http://3-magi.net">cmb</a> to show which members are logged in.<p>'
           . "\n";

    }


// action on choosing "members list" (Plugin main) in plugin menu
//================================================================
    if($admin == 'plugin_main' && $action != 'memberslist') 
    {
        // activation/update
        //==========================================
        if(!$plugin_cf['memberpages']['version']
            || version_compare($plugin_cf['memberpages']['version'], constant('MEMBERPAGES_VERSION'), '!=')) {
            include 'pluginactivation.php';
            $o .= memberpagesUpdate();
        } elseif(isset($_POST['sendPwToNewMember'])) $o .= memberpages_SendEmailToNewMember();
        else $o .= memberpages_ShowMembersList();
    }


    // action on choosing "log" in plugin menu
    //================================================
    if($admin == 'plugin_config' && $action == 'log') $o .= memberpages_Log();


    // action on changing anything in the members list
    //=================================================
    if($admin == 'plugin_main' && $action == 'memberslist' && !isset($_POST['activate_memberpages'])) {
        $o .= memberpages_ProcessList();
    } 


    // action on choosing "config" in plugin menu
    //================================================
    if($admin == 'plugin_config' && $action == 'plugin_edit' || isset($_POST['save_memberpagesconfig'])) {

        if(isset($_POST['save_memberpagesconfig'])) {
            $t = memberpages_SaveConfig();
			$url = CMSIMPLE_URL . '?&memberpages&admin=plugin_config&action=plugin_edit';
			header("Location: $url", true, 303);
			exit;
        }
        $o .= memberpages_Config();
    }  
}
